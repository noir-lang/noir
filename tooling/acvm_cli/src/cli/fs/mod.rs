pub(super) mod inputs;
pub(super) mod witness;
