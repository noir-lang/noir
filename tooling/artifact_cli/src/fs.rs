use std::path::Path;

use color_eyre::eyre::{self, Context};
use noirc_artifacts::{contract::ContractArtifact, program::ProgramArtifact};

/// A parsed JSON build artifact.
#[derive(Debug, Clone)]
pub enum Artifact {
    Program(ProgramArtifact),
    Contract(ContractArtifact),
}

impl Artifact {
    /// Try to parse an artifact as a binary program or a contract
    pub fn read_from_file(path: &Path) -> eyre::Result<Artifact> {
        let json = std::fs::read(path)
            .with_context(|| format!("failed to read artifact from '{}'", path.display()))?;

        let as_program = || serde_json::from_slice::<ProgramArtifact>(&json).map(Artifact::Program);
        let as_contract =
            || serde_json::from_slice::<ContractArtifact>(&json).map(Artifact::Contract);

        Ok(as_program().or_else(|e| as_contract().map_err(|_| e))?)
    }
}
