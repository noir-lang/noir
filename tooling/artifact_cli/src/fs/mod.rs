pub mod artifact;
pub mod inputs;
pub mod witness;
