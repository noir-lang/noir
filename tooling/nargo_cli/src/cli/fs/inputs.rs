use noirc_abi::{
    input_parser::{Format, InputValue},
    Abi, InputMap, MAIN_RETURN_NAME,
};
use std::{collections::BTreeMap, path::Path};

use crate::errors::FilesystemError;

/// Returns the circuit's parameters and its return value, if one exists.
/// # Examples
///
/// ```ignore
/// let (input_map, return_value): (InputMap, Option<InputValue>) =
///   read_inputs_from_file(path, "Verifier", Format::Toml, &abi)?;
/// ```
pub(crate) fn read_inputs_from_file<P: AsRef<Path>>(
    path: P,
    file_name: &str,
    format: Format,
    abi: &Abi,
) -> Result<(InputMap, Option<InputValue>), FilesystemError> {
    if abi.is_empty() {
        return Ok((BTreeMap::new(), None));
    }

    let file_path = path.as_ref().join(file_name).with_extension(format.ext());
    if !file_path.exists() {
        return Err(FilesystemError::MissingTomlFile(file_name.to_owned(), file_path));
    }

    let input_string = std::fs::read_to_string(file_path).unwrap();
    let mut input_map = format.parse(&input_string, abi)?;
    let return_value = input_map.remove(MAIN_RETURN_NAME);

    Ok((input_map, return_value))
}
