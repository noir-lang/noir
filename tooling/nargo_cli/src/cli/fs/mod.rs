use std::{
    fs::File,
    io::Write,
    path::{Path, PathBuf},
};

pub(super) mod inputs;
pub(super) mod program;
pub(super) mod witness;

pub(super) fn create_named_dir(named_dir: &Path, name: &str) -> PathBuf {
    std::fs::create_dir_all(named_dir)
        .unwrap_or_else(|_| panic!("could not create the `{name}` directory"));

    PathBuf::from(named_dir)
}

pub(super) fn write_to_file(bytes: &[u8], path: &Path) -> String {
    let display = path.display();

    let mut file = match File::create(path) {
        Err(why) => panic!("couldn't create {display}: {why}"),
        Ok(file) => file,
    };

    match file.write_all(bytes) {
        Err(why) => panic!("couldn't write to {display}: {why}"),
        Ok(_) => display.to_string(),
    }
}
