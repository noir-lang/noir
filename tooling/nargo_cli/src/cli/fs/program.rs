use std::path::{Path, PathBuf};

use nargo::artifacts::{contract::ContractArtifact, program::ProgramArtifact};
use noirc_frontend::graph::CrateName;

use crate::errors::FilesystemError;

use super::{create_named_dir, write_to_file};

pub(crate) fn save_program_to_file<P: AsRef<Path>>(
    program_artifact: &ProgramArtifact,
    crate_name: &CrateName,
    circuit_dir: P,
) -> PathBuf {
    let circuit_name: String = crate_name.into();
    save_build_artifact_to_file(program_artifact, &circuit_name, circuit_dir)
}

pub(crate) fn save_contract_to_file<P: AsRef<Path>>(
    compiled_contract: &ContractArtifact,
    circuit_name: &str,
    circuit_dir: P,
) -> PathBuf {
    save_build_artifact_to_file(compiled_contract, circuit_name, circuit_dir)
}

fn save_build_artifact_to_file<P: AsRef<Path>, T: ?Sized + serde::Serialize>(
    build_artifact: &T,
    artifact_name: &str,
    circuit_dir: P,
) -> PathBuf {
    create_named_dir(circuit_dir.as_ref(), "target");
    let circuit_path = circuit_dir.as_ref().join(artifact_name).with_extension("json");

    write_to_file(&serde_json::to_vec(build_artifact).unwrap(), &circuit_path);

    circuit_path
}

pub(crate) fn read_program_from_file<P: AsRef<Path>>(
    circuit_path: P,
) -> Result<ProgramArtifact, FilesystemError> {
    let file_path = circuit_path.as_ref().with_extension("json");

    let input_string =
        std::fs::read(&file_path).map_err(|_| FilesystemError::PathNotValid(file_path))?;
    let program = serde_json::from_slice(&input_string)
        .map_err(|err| FilesystemError::ProgramSerializationError(err.to_string()))?;

    Ok(program)
}
