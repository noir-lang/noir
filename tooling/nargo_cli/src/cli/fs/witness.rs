use std::path::{Path, PathBuf};

use acvm::acir::native_types::WitnessStack;
use nargo::constants::WITNESS_EXT;

use super::{create_named_dir, write_to_file};
use crate::errors::FilesystemError;

pub(crate) fn save_witness_to_dir<P: AsRef<Path>>(
    witness_stack: WitnessStack,
    witness_name: &str,
    witness_dir: P,
) -> Result<PathBuf, FilesystemError> {
    create_named_dir(witness_dir.as_ref(), "witness");
    let witness_path = witness_dir.as_ref().join(witness_name).with_extension(WITNESS_EXT);

    let buf: Vec<u8> = witness_stack.try_into()?;

    write_to_file(buf.as_slice(), &witness_path);

    Ok(witness_path)
}
