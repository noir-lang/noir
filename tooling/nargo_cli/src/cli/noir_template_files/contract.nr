contract Main {
    internal fn double(x: Field) -> pub Field { x * 2 }
    fn triple(x: Field) -> pub Field { x * 3 }
    fn quadruple(x: Field) -> pub Field { double(double(x)) }
}
