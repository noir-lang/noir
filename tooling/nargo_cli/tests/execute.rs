#[allow(unused_imports)]
#[cfg(test)]
mod tests {
    use acvm::FieldElement;
    use acvm::acir::circuit::Program;
    // Some of these imports are consumed by the injected tests
    use assert_cmd::prelude::*;
    use base64::Engine;
    use insta::assert_snapshot;
    use insta::internals::Content;
    use insta::internals::Redaction;
    use noirc_artifacts::contract::ContractArtifact;
    use noirc_artifacts::program::ProgramArtifact;
    use predicates::prelude::*;
    use serde::Deserialize;

    use std::collections::BTreeMap;
    use std::fs;
    use std::io::BufWriter;
    use std::path::{Path, PathBuf};
    use std::process::Command;

    use super::*;

    // Utilities to keep the test matrix labels more intuitive.
    #[derive(Debug, Clone, Copy)]
    struct ForceBrillig(pub bool);
    #[derive(Debug, Clone, Copy)]
    struct Inliner(pub i64);

    fn setup_nargo(
        test_program_dir: &Path,
        test_command: &str,
        force_brillig: ForceBrillig,
        inliner_aggressiveness: Inliner,
    ) -> Command {
        let mut nargo = Command::cargo_bin("nargo").unwrap();
        nargo.arg("--program-dir").arg(test_program_dir);
        nargo.arg(test_command).arg("--force");
        nargo.arg("--inliner-aggressiveness").arg(inliner_aggressiveness.0.to_string());
        // Allow more bytecode in exchange to catch illegal states.
        nargo.arg("--enable-brillig-debug-assertions");

        // Enable pedantic solving
        let skip_pedantic_solving = [
            // TODO(https://github.com/noir-lang/noir/issues/8098): all of these are failing with:
            // ```
            // Failed to solve program:
            // \'Failed to solve blackbox function: embedded_curve_add, reason: Infinite input: embedded_curve_add(infinity, infinity)\'
            // ```
            "execution_success/multi_scalar_mul",
            "execution_success/regression_5045",
            "execution_success/regression_7744",
        ];
        if !skip_pedantic_solving
            .into_iter()
            .any(|test_to_skip| test_program_dir.ends_with(test_to_skip))
        {
            nargo.arg("--pedantic-solving");
        }

        // Enable enums and ownership as unstable features
        nargo.arg("-Zenums");

        if force_brillig.0 {
            {
                nargo.arg("--force-brillig");

                // Set the maximum increase so that part of the optimization is exercised (it might fail).
                nargo.arg("--max-bytecode-increase-percent");
                nargo.arg("50");
            }
        }

        nargo
    }

    fn remove_noise_lines(string: String) -> String {
        string
            .lines()
            .filter(|line| {
                !line.contains("Witness saved to")
                    && !line.contains("Circuit witness successfully solved")
                    && !line.contains("Waiting for lock")
            })
            .collect::<Vec<&str>>()
            .join("\n")
    }

    fn delete_test_program_dir_occurrences(string: String, test_program_dir: &Path) -> String {
        // Assuming `test_program_dir` is "/projects/noir/test_programs/compile_failure/some_program"...

        // `test_program_base_dir` is "/projects/noir/test_programs"
        let test_program_base_dir = test_program_dir.parent().unwrap().parent().unwrap();

        // `root_dir` is "/projects/noir"
        let root_dir = test_program_base_dir.parent().unwrap();

        // Here we turn the paths into strings and ensure they end with a `/`
        let mut test_program_base_dir = test_program_base_dir.to_string_lossy().to_string();
        if !test_program_base_dir.ends_with('/') {
            test_program_base_dir.push('/');
        }

        let mut test_program_dir = test_program_dir.to_string_lossy().to_string();
        if !test_program_dir.ends_with('/') {
            test_program_dir.push('/');
        }

        // `test_program_dir_without_root is "test_programs/compile_failure".
        // This one is needed because tests might run from the root of the project and paths
        // will end up starting with "test_programs/compile_failure/...".
        let test_program_dir_without_root =
            test_program_dir.strip_prefix(&root_dir.to_string_lossy().to_string()).unwrap();
        let test_program_dir_without_root = test_program_dir_without_root
            .strip_prefix('/')
            .unwrap_or(test_program_dir_without_root);

        // We replace all of these:
        // - test_program_dir ("/projects/noir/test_programs/compile_failure/foo")
        // - test_program_base_dir ("/projects/noir/test_programs")
        // - test_program_dir_without_root ("test_programs/compile_failure")
        string
            .lines()
            .map(|line| {
                line.replace(&test_program_dir, "")
                    .replace(&test_program_base_dir, "test_programs/")
                    .replace(test_program_dir_without_root, "")
            })
            .collect::<Vec<String>>()
            .join("\n")
    }

    fn execution_success(
        mut nargo: Command,
        test_program_dir: PathBuf,
        check_stdout: bool,
        force_brillig: ForceBrillig,
        inliner: Inliner,
    ) {
        let target_dir = tempfile::tempdir().unwrap().into_path();

        nargo.arg(format!("--target-dir={}", target_dir.to_string_lossy()));

        nargo.assert().success();

        if check_stdout {
            let output = nargo.output().unwrap();
            let stdout = String::from_utf8(output.stdout).unwrap();
            let stdout = remove_noise_lines(stdout);

            let stdout_path = test_program_dir.join("stdout.txt");
            let expected_stdout = if stdout_path.exists() {
                String::from_utf8(fs::read(stdout_path.clone()).unwrap()).unwrap()
            } else {
                String::new()
            };

            // Remove any trailing newlines added by some editors
            let stdout = stdout.trim();
            let expected_stdout = expected_stdout.trim();

            if stdout != expected_stdout {
                if std::env::var("OVERWRITE_TEST_OUTPUT").is_ok() {
                    fs::write(stdout_path, stdout.to_string() + "\n").unwrap();
                } else {
                    println!(
                        "stdout does not match expected output. Expected:\n{expected_stdout}\n\nActual:\n{stdout}"
                    );
                    if expected_stdout.is_empty() && !stdout_path.exists() {
                        println!(
                            "Hint: set the OVERWRITE_TEST_OUTPUT env var to establish a stdout.txt"
                        )
                    }
                    assert_eq!(stdout, expected_stdout);
                }
            }
        }

        check_program_artifact(
            "execution_success",
            &test_program_dir,
            &target_dir,
            force_brillig,
            inliner,
        );
    }

    fn execution_failure(mut nargo: Command) {
        nargo
            .assert()
            .failure()
            .stderr(predicate::str::contains("The application panicked (crashed).").not());
    }

    fn execution_panic(mut nargo: Command) {
        nargo
            .assert()
            .failure()
            .stderr(predicate::str::contains("The application panicked (crashed)."));
    }

    fn noir_test_success(mut nargo: Command) {
        nargo.assert().success();
    }

    fn noir_test_failure(mut nargo: Command) {
        nargo
            .assert()
            .failure()
            .stderr(predicate::str::contains("The application panicked (crashed).").not());
    }

    fn compile_success_empty(
        mut nargo: Command,
        test_program_dir: PathBuf,
        no_warnings: bool,
        force_brillig: ForceBrillig,
        inliner: Inliner,
    ) {
        let target_dir = test_program_dir
            .join(format!("target_force_brillig_{}_inliner_{}", force_brillig.0, inliner.0));

        nargo.arg(format!("--target-dir={}", target_dir.to_string_lossy()));

        nargo.arg("--json");

        let output = nargo.output().expect("Failed to execute command");

        if !output.status.success() {
            {
                panic!(
                    "`nargo info` failed with: {}",
                    String::from_utf8(output.stderr).unwrap_or_default()
                );
            }
        }

        if no_warnings {
            nargo.assert().success().stderr(predicate::str::contains("warning:").not());
        }

        // `compile_success_empty` tests should be able to compile down to an empty circuit.
        let json: serde_json::Value = serde_json::from_slice(&output.stdout).unwrap_or_else(|e| {
            {
                panic!(
                    "JSON was not well-formatted {:?}\n\n{:?}",
                    e,
                    std::str::from_utf8(&output.stdout)
                )
            }
        });
        let num_opcodes = &json["programs"][0]["functions"][0]["opcodes"];
        assert_eq!(
            num_opcodes.as_u64().expect("number of opcodes should fit in a u64"),
            0,
            "expected the number of opcodes to be 0"
        );

        check_program_artifact(
            "compile_success_empty",
            &test_program_dir,
            &target_dir,
            force_brillig,
            inliner,
        );
    }

    fn compile_success_contract(
        mut nargo: Command,
        test_program_dir: PathBuf,
        force_brillig: ForceBrillig,
        inliner: Inliner,
    ) {
        let target_dir = test_program_dir
            .join(format!("target_force_brillig_{}_inliner_{}", force_brillig.0, inliner.0));
        nargo.arg(format!("--target-dir={}", target_dir.to_string_lossy()));

        nargo.assert().success().stderr(predicate::str::contains("warning:").not());

        check_contract_artifact(
            "compile_success_contract",
            &test_program_dir,
            &target_dir,
            force_brillig,
            inliner,
        );
    }

    fn compile_success_no_bug(mut nargo: Command) {
        nargo.assert().success().stderr(predicate::str::contains("bug:").not());
    }

    fn compile_success_with_bug(mut nargo: Command) {
        nargo.assert().success().stderr(predicate::str::contains("bug:"));
    }

    fn compile_failure(mut nargo: Command, test_program_dir: PathBuf) {
        nargo
            .assert()
            .failure()
            .stderr(predicate::str::contains("The application panicked (crashed).").not());

        let output = nargo.output().unwrap();
        let stderr = String::from_utf8(output.stderr).unwrap();
        let stderr = remove_noise_lines(stderr);
        let stderr = delete_test_program_dir_occurrences(stderr, &test_program_dir);

        let stderr_path = test_program_dir.join("stderr.txt");

        let expected_stderr = if stderr_path.exists() {
            String::from_utf8(fs::read(stderr_path.clone()).unwrap()).unwrap()
        } else {
            String::new()
        };

        // Remove any trailing newlines added by some editors
        let stderr = stderr.trim();
        let expected_stderr = expected_stderr.trim();

        if stderr != expected_stderr {
            if std::env::var("OVERWRITE_TEST_OUTPUT").is_ok() {
                fs::write(stderr_path, stderr.to_string() + "\n").unwrap();
            } else {
                // If the expected stderr is empty this is likely a new test, so we produce the expected output for next time
                if expected_stderr.is_empty() {
                    fs::write(stderr_path, stderr.to_string() + "\n").unwrap();
                }

                println!(
                    "stderr does not match expected output. Expected:\n{expected_stderr}\n\nActual:\n{stderr}"
                );
                assert_eq!(stderr, expected_stderr);
            }
        }
    }

    fn nargo_expand_execute(test_program_dir: PathBuf) {
        // First run `nargo execute` on the original code to get the output
        let mut nargo = Command::cargo_bin("nargo").unwrap();
        nargo.arg("--program-dir").arg(test_program_dir.clone());
        nargo.arg("execute").arg("--force").arg("--disable-comptime-printing");

        // Enable enums as an unstable feature
        nargo.arg("-Zenums");

        // Enable pedantic solving
        nargo.arg("--pedantic-solving");

        nargo.assert().success();

        let original_output = nargo.output().unwrap();
        let original_output: String = String::from_utf8(original_output.stdout).unwrap();

        let mut nargo = Command::cargo_bin("nargo").unwrap();
        nargo.arg("--program-dir").arg(test_program_dir.clone());
        nargo.arg("expand").arg("--force").arg("--disable-comptime-printing");

        // Enable enums as an unstable feature
        nargo.arg("-Zenums");

        // Enable pedantic solving
        nargo.arg("--pedantic-solving");

        nargo.assert().success();

        let expanded_code = nargo.output().unwrap();
        let expanded_code: String = String::from_utf8(expanded_code.stdout).unwrap();

        let test_name = test_program_dir.file_name().unwrap().to_string_lossy().to_string();
        let snapshot_name = "expanded";
        insta::with_settings!(
        {
            snapshot_path => format!("./snapshots/expand/execution_success/{test_name}")
        },
        {
            insta::assert_snapshot!(snapshot_name, expanded_code)
        });

        // Create a new directory where we'll put the expanded code
        let temp_dir = tempfile::tempdir().unwrap().into_path();

        // Copy everything from the original directory to the new directory
        // (because some depdendencies might be there and might be needed for the expanded code to work)
        copy_dir_all(test_program_dir.clone(), temp_dir.clone()).unwrap();

        // Create a main file for the expanded code
        fs::write(temp_dir.join("src").join("main.nr"), expanded_code).unwrap();

        // First check if `nargo fmt` works on the expanded code. If not, it means the code is not valid.
        run_nargo_fmt(temp_dir.clone());

        // Now we can run `nargo execute` on the expanded code
        let mut nargo = Command::cargo_bin("nargo").unwrap();
        nargo.arg("--program-dir").arg(temp_dir);
        nargo.arg("execute").arg("--force").arg("--disable-comptime-printing");

        // Enable enums as an unstable feature
        nargo.arg("-Zenums");

        // Enable pedantic solving
        nargo.arg("--pedantic-solving");

        nargo.assert().success();

        let expanded_output = nargo.output().unwrap();
        let expanded_output: String = String::from_utf8(expanded_output.stdout).unwrap();

        let original_output = remove_noise_lines(original_output);
        let expanded_output = remove_noise_lines(expanded_output);

        assert_eq!(original_output, expanded_output);
    }

    fn nargo_expand_compile(test_program_dir: PathBuf, prefix: &'static str) {
        let mut nargo = Command::cargo_bin("nargo").unwrap();
        nargo.arg("--program-dir").arg(test_program_dir.clone());
        nargo.arg("expand").arg("--force").arg("--disable-comptime-printing");

        // Enable enums as an unstable feature
        nargo.arg("-Zenums");

        // Enable pedantic solving
        nargo.arg("--pedantic-solving");

        nargo.assert().success();

        let expanded_code = nargo.output().unwrap();
        let expanded_code: String = String::from_utf8(expanded_code.stdout).unwrap();

        let test_name = test_program_dir.file_name().unwrap().to_string_lossy().to_string();
        let snapshot_name = "expanded";
        insta::with_settings!(
        {
            snapshot_path => format!("./snapshots/expand/{prefix}/{test_name}")
        },
        {
            insta::assert_snapshot!(snapshot_name, expanded_code)
        });

        // Create a new directory where we'll put the expanded code
        let temp_dir = tempfile::tempdir().unwrap().into_path();

        // Copy everything from the original directory to the new directory
        // (because some depdendencies might be there and might be needed for the expanded code to work)
        copy_dir_all(test_program_dir.clone(), temp_dir.clone()).unwrap();

        // Create a main file for the expanded code
        fs::write(temp_dir.join("src").join("main.nr"), expanded_code).unwrap();

        // First check if `nargo fmt` works on the expanded code. If not, it means the code is not valid.
        run_nargo_fmt(temp_dir.clone());

        // Now we can run `nargo compile` on the expanded code
        let mut nargo = Command::cargo_bin("nargo").unwrap();
        nargo.arg("--program-dir").arg(temp_dir);
        nargo.arg("compile").arg("--force");

        // Enable enums as an unstable feature
        nargo.arg("-Zenums");

        // Enable pedantic solving
        nargo.arg("--pedantic-solving");

        nargo.assert().success();
    }

    fn run_nargo_fmt(target_dir: PathBuf) {
        let mut nargo = Command::cargo_bin("nargo").unwrap();
        nargo.arg("--program-dir").arg(target_dir);
        nargo.arg("fmt");
        nargo.assert().success();
    }

    fn check_program_artifact(
        prefix: &'static str,
        test_program_dir: &Path,
        target_dir: &PathBuf,
        force_brillig: ForceBrillig,
        inliner: Inliner,
    ) {
        let artifact_filename =
            find_program_artifact_in_dir(target_dir).expect("Expected an artifact to exist");

        let artifact_file = fs::File::open(&artifact_filename).unwrap();
        let artifact: ProgramArtifact = serde_json::from_reader(artifact_file).unwrap();

        let _ = fs::remove_dir_all(target_dir);

        let test_name = test_program_dir.file_name().unwrap().to_string_lossy().to_string();
        if test_name == "workspace" {
            // workspace outputs multiple artifacts so we get a non-deterministic result.
            return;
        }

        let snapshot_name = format!("force_brillig_{}_inliner_{}", force_brillig.0, inliner.0);
        insta::with_settings!(
            {
                snapshot_path => format!("./snapshots/{prefix}/{test_name}")
            },
            {
            insta::assert_json_snapshot!(snapshot_name, artifact, {
                ".noir_version" => "[noir_version]",
                ".hash" => "[hash]",
                ".bytecode" => insta::dynamic_redaction(|value, _path| {
                    // assert that the value looks like a uuid here
                    let bytecode_b64 = value.as_str().unwrap();
                    let bytecode = base64::engine::general_purpose::STANDARD
                        .decode(bytecode_b64)
                        .unwrap();
                    let program = Program::<FieldElement>::deserialize_program(&bytecode).unwrap();
                    Content::Seq(program.to_string().split("\n").filter(|line: &&str| !line.is_empty()).map(Content::from).collect::<Vec<Content>>())
                }),
                ".file_map.**.path" => file_map_path_redaction(),
            })
        })
    }

    fn check_contract_artifact(
        prefix: &'static str,
        test_program_dir: &Path,
        target_dir: &PathBuf,
        force_brillig: ForceBrillig,
        inliner: Inliner,
    ) {
        let artifact_filename =
            find_program_artifact_in_dir(target_dir).expect("Expected an artifact to exist");

        let artifact_file = fs::File::open(&artifact_filename).unwrap();
        let artifact: ContractArtifact = serde_json::from_reader(artifact_file).unwrap();

        fs::remove_dir_all(target_dir).expect("Could not remove target dir");

        let test_name = test_program_dir.file_name().unwrap().to_string_lossy().to_string();

        let snapshot_name = format!("force_brillig_{}_inliner_{}", force_brillig.0, inliner.0);
        insta::with_settings!(
            {
                snapshot_path => format!("./snapshots/{prefix}/{test_name}")
            },
            {
            insta::assert_json_snapshot!(snapshot_name, artifact, {
                ".noir_version" => "[noir_version]",
                ".functions[].hash" => "[hash]",
                ".file_map.**.path" => file_map_path_redaction(),
            })
        })
    }

    fn file_map_path_redaction() -> Redaction {
        insta::dynamic_redaction(|value, _path| {
            // Some paths are absolute: clear those out.
            let value = value.as_str().expect("Expected a string value in a path entry");
            if value.starts_with("/") { String::new() } else { value.to_string() }
        })
    }

    fn find_program_artifact_in_dir(dir: &PathBuf) -> Option<PathBuf> {
        if !dir.exists() {
            return None;
        }

        for entry in fs::read_dir(dir).unwrap() {
            let Ok(entry) = entry else {
                continue;
            };

            let path = entry.path();
            if path.extension().is_none_or(|ext| ext != "json") {
                continue;
            };

            return Some(path);
        }

        None
    }

    fn copy_dir_all(src: impl AsRef<Path>, dst: impl AsRef<Path>) -> std::io::Result<()> {
        {
            fs::create_dir_all(&dst)?;
            for entry in fs::read_dir(src)? {
                let entry = entry?;
                let ty = entry.file_type()?;
                if ty.is_dir() {
                    copy_dir_all(entry.path(), dst.as_ref().join(entry.file_name()))?;
                } else {
                    fs::copy(entry.path(), dst.as_ref().join(entry.file_name()))?;
                }
            }
            Ok(())
        }
    }

    // include tests generated by `build.rs`
    include!(concat!(env!("OUT_DIR"), "/execute.rs"));
}
