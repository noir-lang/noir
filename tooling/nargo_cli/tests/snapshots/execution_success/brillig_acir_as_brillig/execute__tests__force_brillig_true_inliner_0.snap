---
source: tooling/nargo_cli/tests/execute.rs
expression: artifact
---
{
  "noir_version": "[noir_version]",
  "hash": "[hash]",
  "abi": {
    "parameters": [
      {
        "name": "x",
        "type": {
          "kind": "integer",
          "sign": "unsigned",
          "width": 32
        },
        "visibility": "private"
      }
    ],
    "return_type": null,
    "error_types": {
      "5019202896831570965": {
        "error_kind": "string",
        "string": "attempt to add with overflow"
      },
      "17843811134343075018": {
        "error_kind": "string",
        "string": "Stack too deep"
      }
    }
  },
  "bytecode": [
    "func 0",
    "current witness index : _0",
    "private parameters indices : [_0]",
    "public parameters indices : []",
    "return value indices : []",
    "BRILLIG CALL func 0: inputs: [Single(Expression { mul_terms: [], linear_combinations: [(1, Witness(0))], q_c: 0 })], outputs: []",
    "unconstrained func 0",
    "[Const { destination: Direct(2), bit_size: Integer(U32), value: 1 }, Const { destination: Direct(1), bit_size: Integer(U32), value: 32837 }, Const { destination: Direct(0), bit_size: Integer(U32), value: 3 }, Const { destination: Relative(2), bit_size: Integer(U32), value: 1 }, Const { destination: Relative(3), bit_size: Integer(U32), value: 0 }, CalldataCopy { destination_address: Direct(32836), size_address: Relative(2), offset_address: Relative(3) }, Cast { destination: Direct(32836), source: Direct(32836), bit_size: Integer(U32) }, Mov { destination: Relative(1), source: Direct(32836) }, Call { location: 13 }, Call { location: 14 }, Const { destination: Relative(1), bit_size: Integer(U32), value: 32837 }, Const { destination: Relative(2), bit_size: Integer(U32), value: 0 }, Stop { return_data: HeapVector { pointer: Relative(1), size: Relative(2) } }, Return, Call { location: 30 }, Const { destination: Relative(2), bit_size: Integer(U32), value: 1 }, BinaryIntOp { destination: Relative(3), op: Add, bit_size: U32, lhs: Relative(1), rhs: Relative(2) }, BinaryIntOp { destination: Relative(4), op: LessThanEquals, bit_size: U32, lhs: Relative(1), rhs: Relative(3) }, JumpIf { condition: Relative(4), location: 20 }, Call { location: 36 }, BinaryIntOp { destination: Relative(1), op: Add, bit_size: U32, lhs: Relative(3), rhs: Relative(2) }, BinaryIntOp { destination: Relative(4), op: LessThanEquals, bit_size: U32, lhs: Relative(3), rhs: Relative(1) }, JumpIf { condition: Relative(4), location: 24 }, Call { location: 36 }, Const { destination: Relative(2), bit_size: Integer(U32), value: 2 }, BinaryIntOp { destination: Relative(3), op: Equals, bit_size: U32, lhs: Relative(1), rhs: Relative(2) }, JumpIf { condition: Relative(3), location: 29 }, Const { destination: Relative(4), bit_size: Integer(U32), value: 0 }, Trap { revert_data: HeapVector { pointer: Direct(1), size: Relative(4) } }, Return, Const { destination: Direct(32772), bit_size: Integer(U32), value: 30720 }, BinaryIntOp { destination: Direct(32771), op: LessThan, bit_size: U32, lhs: Direct(0), rhs: Direct(32772) }, JumpIf { condition: Direct(32771), location: 35 }, IndirectConst { destination_pointer: Direct(1), bit_size: Integer(U64), value: 17843811134343075018 }, Trap { revert_data: HeapVector { pointer: Direct(1), size: Direct(2) } }, Return, IndirectConst { destination_pointer: Direct(1), bit_size: Integer(U64), value: 5019202896831570965 }, Trap { revert_data: HeapVector { pointer: Direct(1), size: Direct(2) } }, Return]"
  ],
  "debug_symbols": "nZLBjoQgDED/pWcOgoM4/srEGFSckBA0jGyyMfz7FtFZPbjZzIVnLa8tSRfoVeufjbbD+ILqsUDrtDH62Zixk7MeLf5dIIsH5VAxArRIEAllwn0Fw4s3BE1gCXnCDSqO4AlFgkjAKjwEAnvXZnZKxaaHMXC4STplZ6isN4bAlzR+vfSapF05S4fZjICyPRILDtqo+BXIr51dq0Js7p2+ZX626bXNSrrprCw+8QV/++LKz//wabH77HL+f73+PH2Nkey0Oy1FiHWclq1RWzh42x2y8/e0Z/almtzYqd47FSsdNgvPR56RnNe4SjQGBcnLOsTWPw==",
  "file_map": {
    "50": {
      "source": "fn main(x: u32) {\n    // Safety: testing context\n    unsafe {\n        assert(entry_point(x) == 2);\n        swap_entry_point(x, x + 1);\n        assert(deep_entry_point(x) == 4);\n    }\n}\n\nfn inner(x: u32) -> u32 {\n    x + 1\n}\n\nunconstrained fn entry_point(x: u32) -> u32 {\n    inner(x + 1)\n}\n\nfn swap(x: u32, y: u32) -> (u32, u32) {\n    (y, x)\n}\n\nunconstrained fn swap_entry_point(x: u32, y: u32) {\n    let swapped = swap(x, y);\n    assert(swapped.0 == y);\n    assert(swapped.1 == x);\n    let swapped_twice = swap(swapped.0, swapped.1);\n    assert(swapped_twice.0 == x);\n    assert(swapped_twice.1 == y);\n}\n\nfn level_3(x: u32) -> u32 {\n    x + 1\n}\n\nfn level_2(x: u32) -> u32 {\n    level_3(x + 1)\n}\n\nfn level_1(x: u32) -> u32 {\n    level_2(x + 1)\n}\n\nunconstrained fn deep_entry_point(x: u32) -> u32 {\n    level_1(x + 1)\n}\n",
      "path": ""
    }
  },
  "names": [
    "main"
  ],
  "brillig_names": [
    "main"
  ]
}
